pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py

```diff
@@ -0,0 +1,189 @@
+
+
+class TestData(TypedDict):
+    name: str
+    path: str
+    type_: TestNodeTypeEnum
+    id_: str
+
+
+class TestItem(TestData):
+    lineno: str
+    runID: str
+
+
+class TestNode(TestData):
+    children: "List[TestNode | TestItem]"
+
+
+# Add the path to pythonFiles to sys.path to find testing_tools.socket_manager.
+PYTHON_FILES = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+sys.path.insert(0, PYTHON_FILES)
+
+# Add the lib path to sys.path to find the typing_extensions module.
+sys.path.insert(0, os.path.join(PYTHON_FILES, "lib", "python"))
+from testing_tools import socket_manager
+from typing_extensions import NotRequired
+
+DEFAULT_PORT = "45454"
+
+
+def pytest_collection_finish(session):
+    node, error = build_test_tree(session)
+    cwd = os.getcwd()
+    # TODO: add error checking.
+    sendPost(cwd, node)
+
+
+def build_test_tree(session) -> Tuple[Union[TestNode, None], List[str]]:
+    errors: List[str] = []
+    session_node = create_session_node(session)
+    # a dictionary of all direct children of the session.
+    session_children_dict: dict[str, TestNode] = dict()
+    # a dictionary of all files in the session.
+    file_nodes_dict: dict[pytest.Module, TestNode] = dict()
+    # a dictionary of all classes in the session.
+    class_nodes_dict: dict[str, TestNode] = dict()
+    # iterate through all the test items in the session.
+    for test_case in session.items:
+        test_node = create_test_node(test_case)
+        # Check parent node type, either Module or UnitTest class.
+        if type(test_case.parent) == pytest.Module:
+            file_nodes_dict.setdefault(
+                test_case.parent, create_file_node(test_case.parent)
+            )["children"].append(test_node)
+        else:
+            test_class_node = class_nodes_dict.setdefault(
+                test_case.parent.name,
+                create_class_node(test_case.parent),
+            )
+            test_class_node["children"].append(test_node)
+            parent_module = test_case.parent.parent
+            # Create a file node that has the class as a child.
+            test_file_node = file_nodes_dict.setdefault(
+                parent_module, create_file_node(parent_module)
+            )
+            # Check if the class is already a child of the file node.
+            if test_class_node not in test_file_node["children"]:
+                test_file_node["children"].append(test_class_node)
+
+    created_files_folders_dict: dict[str, TestNode] = {}
+    for file_module, file_node in file_nodes_dict.items():
+        root_folder_node = build_nested_folders(
+            file_module, file_node, created_files_folders_dict, session
+        )
+        # the final folder we get to is the highest folder in the path and therefore we add this as a child to the session.
+        if root_folder_node.get("id_") not in session_children_dict:
+            session_children_dict[root_folder_node.get("id_")] = root_folder_node
```

Comment:
Should `None` be used as a key in `session_children_dict`, as `dict.get()` will return that if it isn't there? And what if "id_" isn't there? Is `None` a legitimate key?

And assuming `None` is what you want if `"id_"` doesn't exist, then I would not get it twice:

```suggestion
        root_folder_id = root_folder_node.get("id_")
        if root_folder_id not in session_children_dict:
            session_children_dict[root_folder_id] = root_folder_node
```